Update testharness.py

jrper opened the pull request:
Apply the change proposed in #182 to allow oversubscription of MPI slots in openmpi 3.0 and above.

jrper commented:
I've added the parallel test harness code to this to put that into a PR. It only gives a half hour speed up on Jenkins, but I think that's still useful.

tmbgreaves commented:
I'll have a go now at specifying a much heftier VM and see if that gives significant benefits - will report back here..

tmbgreaves commented:
5x speed-up going from 2 threads to 16 threads for short tests.

tmbgreaves commented:
Also trying 32-core, still on D-type nodes, and 64-core on compute-optimised F64s_v2

jrper commented:
How much should we be valuing the machine efficiency? Speaking personally, there's probably not much difference between a queue taking one or two hours, but 4 hours is half a developer day, and the current ten hours means "come back tomorrow".

tmbgreaves commented:
I think we want to get this down under 2h if possible, giving the opportunity of multiple test-hack-retest cycles in a day. So far the 32-core and 64-core runs aren't showing significant (if any) improvement over 16-core, and assuming that the 16-core continues the 4-5x speed-up it's going to come in around the 2h mark.

jrper commented:
I guess that's not all that surprising, given the number of possible bottlenecks this all has (docker+python subprocess parallelism+io). There are certainly a few places where low hanging fruit still exists on the `testharness.py` end (e.g. parallelising the actual python testing, and doing a better job on the io to stdout), and I never even looked at `unittestharness.py`, but a 2 hour test cycle on 16 cores, sounds good to me, assuming we're going to move away from testing every single commit automatically. I'll add @stephankramer as a reviewer for the actual code changes, @tmbgreaves shout when you've settled on a final Jenkinsfile setup.

tmbgreaves commented:
I'd like to check that the 8 doesn't turn out to be roughly the same time as the 16 - I've started that going now.

tmbgreaves commented:
Unittesting comes out on the order of 10 minutes so not the worst offender by some margin.

tmbgreaves commented:
32-core comes in about 15 minutes quicker than 16-core, at 1h25 in the actual running (plus spin-up time).

tmbgreaves commented:
8 cores comes in pretty close to the 16 core - a fraction over 2h on the CentOS 7 run which has completed. 4-core projects out to being about twice that. So I think the sweet spot for now is 8 cores.

jrper commented:
That makes sense, given that I'd only tested locally out to 8 threads and was seeing roughly that order of speed-up. My sense is that the limiting factor is now some deeply embedded python locks controlling the memory that's getting shared between subprocesses, but I'm only guessing and fixing that would take a  very heavy rewrite, to the point where we'd be spinning off our own multi language testing framework. Over to @stephankramer, or whoever else wants to review this. Note, it should probably be rebased when committed to master, since git seems to think it's spuriously diverged.

tmbgreaves commented:
Rebased on master.

stephankramer commented on tools/testharness.py:
```diff
@@ -285,14 +289,41 @@ def run(self):
+                   queue = Queue.Queue()
+                   for test in tests:
+                       queue.put(test)
+                   self.threadrun(queue)
```
An empty line would help here

jrper commented on python/fluidity/regressiontest.py:
```diff
@@ -222,6 +226,10 @@ def Trim(string):
+                sys.stdout = _
+                log.seek(0)
+                tc.stdout = log.read()
+                print tc.stdout
```
Funnily enough, I ran into that trying to update the python3 fixes to work with this. You're right that it ought to be sorted here though.

jrper commented on python/fluidity/regressiontest.py:
```diff
@@ -9,6 +9,7 @@
 import glob
 import threading
 import traceback
+import StringIO as io
```
I'll double check what I did in the python3 compatible branch and backport as necessary.

jrper commented:
@stephankramer Hopefully I've addressed your comments in the latest commit, possibly excluding the full hardening towards python3, which I'm intending to wrap into my next target, #180.
